src/services/refactors/convertToOptionalChainExpression.ts

```diff
@@ -0,0 +1,126 @@
+
+    function getAvailableActions(context: RefactorContext): readonly ApplicableRefactorInfo[] {
+        const info = getInfo(context, context.triggerReason === "invoked");
+        if (!info) return emptyArray;
+        return [{
+            name: refactorName,
+            description: convertToOptionalChainExpressionMessage,
+            actions: [{
+                name: refactorName,
+                description: convertToOptionalChainExpressionMessage
+            }]
+        }];
+    }
+
+    function getEditsForAction(context: RefactorContext, actionName: string): RefactorEditInfo | undefined {
+        const info = getInfo(context);
+        if (!info) return undefined;
+        const edits = textChanges.ChangeTracker.with(context, t => doChange(context.file, context.program.getTypeChecker(), t, info, actionName));
+        return { edits, renameFilename: undefined, renameLocation: undefined };
+    }
+
+    interface Info {
+        fullPropertyAccess: PropertyAccessExpression,
+        firstOccurrence: Node,
+        expression: BinaryExpression | ConditionalExpression
+    }
+
+    function getInfo(context: RefactorContext, considerEmptySpans = true): Info | undefined {
+        const { file, program } = context;
+        const span = getRefactorContextSpan(context);
+
+        if (span.length === 0 && !considerEmptySpans) return undefined;
+        const forEmptySpan = span.length === 0 && considerEmptySpans;
+
+        const startToken = getTokenAtPosition(file, span.start);
+
+        const containingNode = forEmptySpan ? findAncestor(startToken, (node) => { return isExpressionStatement(node) && (isBinaryExpression(node.expression) || isConditionalExpression(node.expression)); }) : getParentNodeInSpan(startToken, file, span);
+        const expression = containingNode && isExpressionStatement(containingNode) ? containingNode.expression : containingNode;
+        if (!expression) return undefined;
+
+        const checker = program.getTypeChecker();
+
+
+        if (isBinaryExpression(expression)) {
+            const start = forEmptySpan ? expression.pos : startToken.pos;
+
+            const fullPropertyAccess = getFullPropertyAccessChain(expression);
+            if (!fullPropertyAccess) return undefined;
+            if (expression.operatorToken.kind !== SyntaxKind.AmpersandAmpersandToken && expression.operatorToken.pos <= fullPropertyAccess.pos) return undefined;
+
+            // ensure that each sequential operand in range matches the longest acceess chain
+            let checkNode = expression.left;
+            let firstOccurrence: PropertyAccessExpression | Identifier = fullPropertyAccess;
+            while (isBinaryExpression(checkNode) && checkNode.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken && (isPropertyAccessExpression(checkNode.right) || isIdentifier(checkNode.right)) && checkNode.right.pos >= start) {
+                if (!checker.containsMatchingReference(fullPropertyAccess, checkNode.right)) {

```

Comment:
I need to spend more time looking at this, but my initial understanding is that fullPropertyAccess=a.b.c.d, for example, and checkNode.right walks along a case like `a && a.b && a.b.c && a.b.c.d`. And since fullPropertyAccess isn't calculated recursively, but containsMatchingReference walks along it recursively, I think my understanding is broken -- backward somehow probably.

Reply:
Yep, `&&` and Property access are both left-associative and parse left-deep, so recursion isn't needed to get the last item, and a linear walk goes backward.

Reply:
switched to using `isMatchingExpression`